[LoRA] Add LoRA support to AuraFlow

hameerabbasi opened the pull request:
# What does this PR do?
This PR is a simple rebase of #9017

cc @sayakpaul for review.

<!--
Congratulations! You've made it this far! You're not quite done yet though.

Once merged, your PR is going to appear in the release notes with the title you set, so make sure it's a great title that fully reflects the extent of your awesome contribution.

Then, please replace this with a description of the change and which issue is fixed (if applicable). Please also include relevant motivation and context. List any dependencies (if any) that are required for this change.

Once you're done, someone will review your PR shortly (see the section "Who can review?" below to tag some potential reviewers). They may suggest changes to make the code even better. If no one reviewed your PR after a week has passed, don't hesitate to post a new comment @-mentioning the same persons---sometimes notifications get lost.
-->

<!-- Remove if not applicable -->

Fixes # (issue)


## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [ ] Did you read the [contributor guideline](https://github.com/huggingface/diffusers/blob/main/CONTRIBUTING.md)?
- [ ] Did you read our [philosophy doc](https://github.com/huggingface/diffusers/blob/main/PHILOSOPHY.md) (important for complex PRs)?
- [ ] Was this discussed/approved via a GitHub issue or the [forum](https://discuss.huggingface.co/c/discussion-related-to-httpsgithubcomhuggingfacediffusers/63)? Please add a link to it if that's the case.
- [ ] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/diffusers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/diffusers/tree/main/docs#writing-source-documentation).
- [ ] Did you write any new necessary tests?


## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

<!-- Your PR will be replied to more quickly if you can figure out the right person to tag with @.

 If you know how to use git blame, that is the easiest way, otherwise, here is a rough guide of **who to tag**.
 Please tag fewer than 3 people.

Core library:

- Schedulers: @yiyixuxu
- Pipelines and pipeline callbacks: @yiyixuxu and @asomoza
- Training examples: @sayakpaul
- Docs: @stevhliu and @sayakpaul
- JAX and MPS: @pcuenca
- Audio: @sanchit-gandhi
- General functionalities: @sayakpaul @yiyixuxu @DN6

Integrations:

- deepspeed: HF Trainer/Accelerate: @SunMarc
- PEFT: @sayakpaul @BenjaminBossan

HF projects:

- accelerate: [different repo](https://github.com/huggingface/accelerate)
- datasets: [different repo](https://github.com/huggingface/datasets)
- transformers: [different repo](https://github.com/huggingface/transformers)
- safetensors: [different repo](https://github.com/huggingface/safetensors)

-->

hameerabbasi commented:
Thanks for the helping hand, @hlky!

hlky commented:
See https://github.com/huggingface/diffusers/blob/main/tests/lora/test_lora_layers_flux.py https://github.com/huggingface/diffusers/blob/main/tests/lora/test_lora_layers_mochi.py etc as an example for tests. Seems to be missing `transformer_cls`, `get_dummy_inputs` and probably others, ping @sayakpaul if you need help with lora tests.

hameerabbasi commented:
@sayakpaul Okay; I'm at a point where I've got actual, valid test failures but have no idea where to look.

hameerabbasi commented:
Here's the log after the latest commit: [pytest.log](https://github.com/user-attachments/files/18139060/pytest.log)

sayakpaul commented on src/diffusers/loaders/lora_pipeline.py:
```diff
@@ -1641,6 +1641,495 @@ def unfuse_lora(self, components: List[str] = ["transformer", "text_encoder", "t
+    Load LoRA layers into [`AuraFlowTransformer2DModel`] Specific to [`AuraFlowPipeline`].
+    """
+
+    _lora_loadable_modules = ["transformer", "text_encoder"]
+    transformer_name = TRANSFORMER_NAME
```
```suggestion
    _lora_loadable_modules = ["transformer", "text_encoder"]
    transformer_name = TRANSFORMER_NAME
    text_encoder_name = TEXT_ENCODER_NAME
```

hameerabbasi replied:
Done.

sayakpaul commented on src/diffusers/loaders/lora_pipeline.py:
```diff
@@ -1641,6 +1641,495 @@ def unfuse_lora(self, components: List[str] = ["transformer", "text_encoder", "t
+    @classmethod
+    @validate_hf_hub_args
+    # Copied from diffusers.loaders.lora_pipeline.StableDiffusionLoraLoaderMixin.lora_state_dict
+    def lora_state_dict(
```
I think we don't think we need to copy-paste it from `StableDiffusionLoraLoaderMixin`. 

It should be copy-pasted from 
https://github.com/huggingface/diffusers/blob/a5f35ee4731b731d6bd8977525873b0bc480cb42/src/diffusers/loaders/lora_pipeline.py#L1149

hameerabbasi replied:
This leads to the following test failure:

```
FAILED tests/lora/test_lora_layers_af.py::AuraFlowLoRATests::test_unexpected_keys_warning - ValueError: too many values to unpack (expected 2)
```

sayakpaul replied:
Please state the full strace. It's difficult to gauge what's happening from a single failure.

hameerabbasi replied:
I got this passing by copying most things from Flux.

sayakpaul commented on src/diffusers/loaders/lora_pipeline.py:
```diff
@@ -1641,6 +1641,495 @@ def unfuse_lora(self, components: List[str] = ["transformer", "text_encoder", "t
+        # then the `state_dict` keys should have `self.unet_name` and/or `self.text_encoder_name` as
+        # their prefixes.
+        keys = list(state_dict.keys())
+        prefix = cls.text_encoder_name if prefix is None else prefix
```
We're using `cls.text_encoder_name` without actually having the attribute `text_encoder_name`. But with the review suggestions, this should be fixed.

hameerabbasi replied:
Done.

sayakpaul commented on src/diffusers/loaders/lora_pipeline.py:
```diff
@@ -1641,6 +1641,495 @@ def unfuse_lora(self, components: List[str] = ["transformer", "text_encoder", "t
+                # Unsafe code />
+
+    @classmethod
+    def save_lora_weights(
```
We're supporting both `transformer` and `text_encoder` as loadable components. So, we should support serializing the text encoder layers, too. 

Feel free to reuse:
https://github.com/huggingface/diffusers/blob/a5f35ee4731b731d6bd8977525873b0bc480cb42/src/diffusers/loaders/lora_pipeline.py#L2148

hameerabbasi replied:
Done.

sayakpaul commented on src/diffusers/loaders/lora_pipeline.py:
```diff
@@ -1641,6 +1641,495 @@ def unfuse_lora(self, components: List[str] = ["transformer", "text_encoder", "t
+            safe_serialization=safe_serialization,
+        )
+
+    def fuse_lora(
```
Should be accompanied with a `# Copied from ...` statement.

hameerabbasi replied:
This was in the previous PR; do you know where it was copied from?

sayakpaul commented on src/diffusers/loaders/lora_pipeline.py:
```diff
@@ -1641,6 +1641,495 @@ def unfuse_lora(self, components: List[str] = ["transformer", "text_encoder", "t
+            components=components, lora_scale=lora_scale, safe_fusing=safe_fusing, adapter_names=adapter_names
+        )
+
+    def unfuse_lora(self, components: List[str] = ["transformer"], **kwargs):
```
Same as above.

sayakpaul reviewed (changes requested):
Thanks for the PR. I have left some comments to fix a couple of things. LMK if they're unclear.

hameerabbasi commented:
Latest [test log](https://github.com/user-attachments/files/18195414/pytest.log).

hameerabbasi commented:
@sayakpaul I've uploaded the latest test log in [this comment](https://github.com/huggingface/diffusers/pull/10216#issuecomment-2553182331), and commented the difficulties I faced with certain changes. Some pointers would be appreciated.

sayakpaul commented on src/diffusers/loaders/lora_pipeline.py:
```diff
@@ -1641,6 +1641,487 @@ def unfuse_lora(self, components: List[str] = ["transformer", "text_encoder", "t
+
+    def fuse_lora(
+        self,
+        components: List[str] = ["transformer"],
```
So, this also should have "text_encoder" since we're including `_lora_loadable_modules = ["transformer", "text_encoder"]`.

And we can include a copy statement from: 
https://github.com/huggingface/diffusers/blob/1896b1f7c1c740648cf163c82efdce5c2c861207/src/diffusers/loaders/lora_pipeline.py#L2204

hameerabbasi replied:
Done in https://github.com/huggingface/diffusers/pull/10216/commits/6514892fd53f06502329b3927714f7d166db89fd

sayakpaul commented on src/diffusers/models/transformers/auraflow_transformer_2d.py:
```diff
@@ -451,6 +452,7 @@ def forward(
         hidden_states: torch.FloatTensor,
         encoder_hidden_states: torch.FloatTensor = None,
         timestep: torch.LongTensor = None,
+        joint_attention_kwargs: Optional[Dict[str, Any]] = None,
```
Let's rename this to `attention_kwargs`.

hameerabbasi replied:
Done in https://github.com/huggingface/diffusers/pull/10216/commits/6514892fd53f06502329b3927714f7d166db89fd

sayakpaul commented on src/diffusers/pipelines/aura_flow/pipeline_aura_flow.py:
```diff
@@ -408,6 +428,7 @@ def __call__(
         max_sequence_length: int = 256,
         output_type: Optional[str] = "pil",
         return_dict: bool = True,
+        joint_attention_kwargs: Optional[Dict[str, Any]] = None,
```
Let's rename this to `attention_kwargs`.

hameerabbasi replied:
Done in https://github.com/huggingface/diffusers/pull/10216/commits/6514892fd53f06502329b3927714f7d166db89fd

hameerabbasi replied:
Copying from Flux would break this -- I'd prefer to keep as-is. Let me know if you'd still like to change it.

hameerabbasi replied:
Actually, done in https://github.com/huggingface/diffusers/pull/10216/commits/d6028cdc53d6d30c14022f8045f87c7f59c466c1.

sayakpaul commented on tests/lora/test_lora_layers_af.py:
```diff
@@ -0,0 +1,114 @@
+# coding=utf-8
```
Let's rename this file to `test_lora_layers_auraflow.py`.

hameerabbasi replied:
Done in https://github.com/huggingface/diffusers/pull/10216/commits/6514892fd53f06502329b3927714f7d166db89fd

hameerabbasi commented:
Here's the latest test log: [pytest.log](https://github.com/user-attachments/files/18327057/pytest.log)

hameerabbasi commented:
[Current pytest.log](https://github.com/user-attachments/files/18327514/pytest.log)

Failures:
```
=========================== short test summary info ============================
FAILED tests/lora/test_lora_layers_auraflow.py::AuraFlowLoRATests::test_correct_lora_configs_with_different_ranks - AssertionError: False is not true
FAILED tests/lora/test_lora_layers_auraflow.py::AuraFlowLoRATests::test_lora_B_bias - AssertionError: True is not false
FAILED tests/lora/test_lora_layers_auraflow.py::AuraFlowLoRATests::test_low_cpu_mem_usage_with_loading - AttributeError: 'AuraFlowPipeline' object has no attribute 'lora_scale'
FAILED tests/lora/test_lora_layers_auraflow.py::AuraFlowLoRATests::test_set_adapters_match_attention_kwargs - AssertionError: True is not false : Lora + scale should change the output
FAILED tests/lora/test_lora_layers_auraflow.py::AuraFlowLoRATests::test_simple_inference_with_dora - AssertionError: True is not false : DoRA lora should change the output
FAILED tests/lora/test_lora_layers_auraflow.py::AuraFlowLoRATests::test_simple_inference_with_partial_text_lora - ValueError: Target modules {'out_proj', 'v_proj', 'q_proj', 'k_proj'} not found in the base model. Please check the target modules and try again.
FAILED tests/lora/test_lora_layers_auraflow.py::AuraFlowLoRATests::test_simple_inference_with_text_denoiser_block_scale - AssertionError: True is not false : LoRA weights 1 and 2 should give different results
FAILED tests/lora/test_lora_layers_auraflow.py::AuraFlowLoRATests::test_simple_inference_with_text_denoiser_lora_and_scale - AssertionError: False is not true : Lora should change the output
FAILED tests/lora/test_lora_layers_auraflow.py::AuraFlowLoRATests::test_simple_inference_with_text_denoiser_lora_save_load - AttributeError: 'AuraFlowPipeline' object has no attribute 'lora_scale'
FAILED tests/lora/test_lora_layers_auraflow.py::AuraFlowLoRATests::test_simple_inference_with_text_denoiser_multi_adapter - AssertionError: True is not false : Adapter outputs should be different.
FAILED tests/lora/test_lora_layers_auraflow.py::AuraFlowLoRATests::test_simple_inference_with_text_denoiser_multi_adapter_block_lora - AssertionError: True is not false : Adapter 1 and 2 should give different results
FAILED tests/lora/test_lora_layers_auraflow.py::AuraFlowLoRATests::test_simple_inference_with_text_denoiser_multi_adapter_delete_adapter - AssertionError: True is not false : Adapter 1 and 2 should give different results
FAILED tests/lora/test_lora_layers_auraflow.py::AuraFlowLoRATests::test_simple_inference_with_text_denoiser_multi_adapter_weighted - AssertionError: True is not false : Adapter 1 and 2 should give different results
FAILED tests/lora/test_lora_layers_auraflow.py::AuraFlowLoRATests::test_simple_inference_with_text_lora - AssertionError: False is not true : Lora should change the output
FAILED tests/lora/test_lora_layers_auraflow.py::AuraFlowLoRATests::test_simple_inference_with_text_lora_and_scale - AssertionError: False is not true : Lora should change the output
FAILED tests/lora/test_lora_layers_auraflow.py::AuraFlowLoRATests::test_simple_inference_with_text_lora_denoiser_fused - AssertionError: True is not false : Fused lora should change the output
FAILED tests/lora/test_lora_layers_auraflow.py::AuraFlowLoRATests::test_simple_inference_with_text_lora_fused - AssertionError: True is not false : Fused lora should change the output
FAILED tests/lora/test_lora_layers_auraflow.py::AuraFlowLoRATests::test_simple_inference_with_text_lora_save_load - AttributeError: 'AuraFlowPipeline' object has no attribute 'lora_scale'
===== 18 failed, 12 passed, 4 skipped, 397 deselected, 1 warning in 23.58s =====
```

hameerabbasi commented:
It seems nothing but CLIP is supported in https://github.com/huggingface/diffusers/blob/532013f990cafd646d340f58ba599b9ace91d241/src/diffusers/models/lora.py#L41-L66

which is called here
https://github.com/huggingface/diffusers/blob/532013f990cafd646d340f58ba599b9ace91d241/src/diffusers/loaders/lora_pipeline.py#L2185-L2197

This means essentially that we need more plumbing to support this for arbitrary text encoders, or to only support the transformer for AuraFlow. This is because AuraFlow only has one UMT5 encoder.

sayakpaul commented:
Thanks, let's support only `transformer` for the moment, then?

sayakpaul commented:
Could we try to look into each failure and debug?

hameerabbasi commented on src/diffusers/loaders/lora_base.py:
```diff
@@ -356,21 +356,13 @@ def _load_lora_into_text_encoder(
-                    if rank_key not in text_encoder_lora_state_dict:
-                        continue
-                    rank[rank_key] = text_encoder_lora_state_dict[rank_key].shape[1]
+            text_encoder_lora_state_dict = convert_state_dict_to_peft(text_encoder_lora_state_dict, original_type=StateDictType.DIFFUSERS)
+
+            for name, module in text_encoder.named_modules():
+                if "lora_A" not in name and "lora_B" not in name and isinstance(module, (nn.Linear, nn.Conv2d)):
+                    rank_key = f"{name.removesuffix(".base_layer")}.lora_B.weight"
+                    if rank_key in text_encoder_lora_state_dict:
+                        rank[rank_key] = text_encoder_lora_state_dict[rank_key].shape[1]
```
These changes support TE LoRA other than CLIP; but they cause one other model to fail, likely an edge case.

sayakpaul replied:
I think it's fine to not add these changes in this PR as it seems unrelated.

hameerabbasi replied:
I'll make a separate PR for this and we can review that one first -- it seems to be the crux of what's failing with the tests IIUC. Then coming back to this one should be easy.

hameerabbasi replied:
It still causes the failures here, and I'm kinda clueless how to fix them: https://github.com/huggingface/diffusers/pull/10216#issuecomment-2577850739

sayakpaul replied:
I don't think so.

https://github.com/huggingface/diffusers/blob/edb8c1bce67e81f0de90a7e4c16b2f6537d39f2d/src/diffusers/loaders/lora_pipeline.py#L2560 doesn't have text encoder either.

bghira replied:
neither does OmniGen

hameerabbasi commented:
Okay, I've updated the code to mirror Mochi1 instead of Flux, as that was a closer match (no TE LoRA), and removed the TE from loadable modules. I've also skipped TE-only LoRA tests. Here are the remaining failures:

<details><summary>Failure list</summary>

```
=========================== short test summary info ============================
FAILED tests/lora/test_lora_layers_auraflow.py::AuraFlowLoRATests::test_correct_lora_configs_with_different_ranks - AssertionError: False is not true
FAILED tests/lora/test_lora_layers_auraflow.py::AuraFlowLoRATests::test_lora_B_bias - AssertionError: True is not false
FAILED tests/lora/test_lora_layers_auraflow.py::AuraFlowLoRATests::test_set_adapters_match_attention_kwargs - AssertionError: True is not false : Lora + scale should change the output
FAILED tests/lora/test_lora_layers_auraflow.py::AuraFlowLoRATests::test_simple_inference_with_dora - AssertionError: True is not false : DoRA lora should change the output
FAILED tests/lora/test_lora_layers_auraflow.py::AuraFlowLoRATests::test_simple_inference_with_text_denoiser_lora_and_scale - AssertionError: False is not true : Lora should change the output
FAILED tests/lora/test_lora_layers_auraflow.py::AuraFlowLoRATests::test_simple_inference_with_text_denoiser_multi_adapter - AssertionError: True is not false : Adapter outputs should be different.
FAILED tests/lora/test_lora_layers_auraflow.py::AuraFlowLoRATests::test_simple_inference_with_text_denoiser_multi_adapter_block_lora - AssertionError: True is not false : Adapter 1 and 2 should give different results
FAILED tests/lora/test_lora_layers_auraflow.py::AuraFlowLoRATests::test_simple_inference_with_text_denoiser_multi_adapter_delete_adapter - AssertionError: True is not false : Adapter 1 and 2 should give different results
FAILED tests/lora/test_lora_layers_auraflow.py::AuraFlowLoRATests::test_simple_inference_with_text_denoiser_multi_adapter_weighted - AssertionError: True is not false : Adapter 1 and 2 should give different results
FAILED tests/lora/test_lora_layers_auraflow.py::AuraFlowLoRATests::test_simple_inference_with_text_lora_denoiser_fused - AssertionError: True is not false : Fused lora should change the output
==== 10 failed, 14 passed, 10 skipped, 398 deselected, 4 warnings in 20.21s ====
```
</details> 

Latest [pytest.log](https://github.com/user-attachments/files/18469191/pytest.log)

sayakpaul commented:
I think a better approach to debugging is taking a single test failure and trying to see what's causing them. We could compare the implementation with another model having similarities (Mochi-1 is a good one) and then take things from there.

Have we tried that? 

We're grateful for your contributions so far. But it might be even better if we tried going further a bit for debugging. This will help solidify your contributions, too. `SANALoRALoaderMixin` is another class where we could take references from:

https://github.com/huggingface/diffusers/blob/328e0d20a7b996f9bdb04180457eb08c1b42a76e/src/diffusers/loaders/lora_pipeline.py#L3199

hameerabbasi commented:
@sayakpaul Okay, I debugged this and it seems like anyplace an adapter is deleted and then added again, or replaced, it doesn't seem to go through.

For example, when debugging `tests/lora/test_lora_layers_auraflow.py::AuraFlowLoRATests::test_correct_lora_configs_with_different_ranks`:

1. Original `pipe` works.
2. Original `adapter-1` works.
3. Removal of `adapter-1` (maybe?) works.
4. Adding new `adapter-1` appears to work by inspecting `pipe.transformer.peft_config["adapter-1"].rank_pattern`, but produces the same results as the original model without a LoRA.

This pattern shows up in all of the failed tests: All of them are applying multiple LoRAs either at the same time or in-sequence.

Would you have an idea where a bug like this could originate?

sayakpaul commented:
Thanks for your work thus far.

We recently added LoRA support to Lumina2 in https://github.com/huggingface/diffusers/pull/10818 which should resemble the changes needed in this PR. Possible to maybe compare and see what are we missing?

hameerabbasi commented:
I've mirrored the pattern in Lumina2, the only difference that could possibly have made a difference was the order of the mixins. However, I corrected that as well.

sayakpaul commented on src/diffusers/pipelines/aura_flow/pipeline_aura_flow.py:
```diff
@@ -634,7 +669,16 @@ def __call__(
 
         return ImagePipelineOutput(images=image)
+
+    @property
+    def attention_kwargs(self):
+        return self._attention_kwargs
```
This should be positioned upper.

hameerabbasi replied:
Done in https://github.com/huggingface/diffusers/pull/10216/commits/3b9e65543a080ab4eb978fe7c9c7f4d0198b1139

hameerabbasi commented on src/diffusers/pipelines/aura_flow/pipeline_aura_flow.py:
```diff
@@ -634,7 +669,16 @@ def __call__(
+        if self.text_encoder is not None:
+            if isinstance(self, AuraFlowLoraLoaderMixin) and USE_PEFT_BACKEND:
+                # Retrieve the original scale by scaling back the LoRA layers
+                unscale_lora_layers(self.text_encoder, lora_scale)
```
Done in https://github.com/huggingface/diffusers/pull/10216/commits/3b9e65543a080ab4eb978fe7c9c7f4d0198b1139

hameerabbasi commented:
It seems the `scale`/`unscale` stuff was needed both in the pipeline and the transformer. I made sure it's there, appropriately spelled, in both. However; this still wasn't enough to get the tests to pass.

sayakpaul commented:
@bot /style

hameerabbasi commented:
@sayakpaul All tests passing and ready for review.

hameerabbasi commented:
@cloneofsimo I'm not entirely sure if https://github.com/huggingface/diffusers/commit/dee9074b9e5a361314e393341bdb827dd6c79619 is the right fix -- would you happen to know if these are the right layers to target?

cloneofsimo commented:
@hameerabbasi Yes, I think these are right layers to target!

hameerabbasi commented:
@sayakpaul Gentle ping, this one is ready for review/merge.

sayakpaul commented:
@hameerabbasi genuinely sorry for the delay. I have started the CI, let's see how it goes.

sayakpaul commented:
@bot /style

hameerabbasi commented:
@sayakpaul Style CI push didn't trigger the full CI.

sayakpaul commented:
@hameerabbasi we have to fix the consistency issues.

hameerabbasi commented:
@sayakpaul Done! Thanks for the pointer.

sayakpaul commented on tests/lora/utils.py:
```diff
@@ -2187,6 +2188,17 @@ def test_correct_lora_configs_with_different_ranks(self):
         self.assertTrue(not np.allclose(original_output, lora_output_diff_alpha, atol=1e-3, rtol=1e-3))
         self.assertTrue(not np.allclose(lora_output_diff_alpha, lora_output_same_rank, atol=1e-3, rtol=1e-3))
 
+    @property
+    def supports_text_encoder_lora(self):
+        return (
+            len(
+                {"text_encoder", "text_encoder_2", "text_encoder_3"}.intersection(
+                    self.pipeline_class._lora_loadable_modules
+                )
+            )
+            != 0
+        )
```
I guess this is not being used? So, maybe we can remove?

hameerabbasi replied:
Done in https://github.com/huggingface/diffusers/pull/10216/commits/e97a83ea7ed53b1ed27834bcbe390bbbc6ab28b3.

sayakpaul commented on tests/lora/utils.py:
```diff
@@ -2239,11 +2251,15 @@ def initialize_pipeline(storage_dtype=None, compute_dtype=torch.float32):
-        pipe_float8_e4m3_fp32(**inputs, generator=torch.manual_seed(0))[0]
+        # MPS doesn't support float8 yet.
+        if torch_device not in {"mps"}:
+            pipe_float8_e4m3_fp32 = initialize_pipeline(storage_dtype=torch.float8_e4m3fn, compute_dtype=torch.float32)
+            pipe_float8_e4m3_fp32(**inputs, generator=torch.manual_seed(0))[0]
 
-        pipe_float8_e4m3_bf16 = initialize_pipeline(storage_dtype=torch.float8_e4m3fn, compute_dtype=torch.bfloat16)
-        pipe_float8_e4m3_bf16(**inputs, generator=torch.manual_seed(0))[0]
+            pipe_float8_e4m3_bf16 = initialize_pipeline(
+                storage_dtype=torch.float8_e4m3fn, compute_dtype=torch.bfloat16
+            )
+            pipe_float8_e4m3_bf16(**inputs, generator=torch.manual_seed(0))[0]
```
This is unrelated. So, please open a new PR for this.

hameerabbasi replied:
Done in https://github.com/huggingface/diffusers/pull/10216/commits/e97a83ea7ed53b1ed27834bcbe390bbbc6ab28b3.

sayakpaul commented on tests/lora/utils.py:
```diff
@@ -2268,7 +2284,7 @@ def test_layerwise_casting_peft_input_autocast_denoiser(self):
         )
 
-        storage_dtype = torch.float8_e4m3fn
+        storage_dtype = torch.float8_e4m3fn if not torch_device == "mps" else torch.bfloat16
```
Same.

hameerabbasi replied:
Done in https://github.com/huggingface/diffusers/pull/10216/commits/e97a83ea7ed53b1ed27834bcbe390bbbc6ab28b3.

sayakpaul reviewed (approved):
This is looking quite nice. Thanks for the cool changes in the test suite as well! I have some last comments. After those are resolved, we should be good to go!

Thanks very much for your patience and apologies for the delay on my side.

hameerabbasi commented:
@sayakpaul CI needs approval again due to https://github.com/huggingface/diffusers/pull/10216/commits/a5b78d166f246789f343d638191b0cc5da322402, sorry.

sayakpaul commented:
@hameerabbasi we have got a failed test here:
https://github.com/huggingface/diffusers/actions/runs/14325947574/job/40151562356?pr=10216#step:7:1678

hameerabbasi commented:
@sayakpaul Right, fixed in https://github.com/huggingface/diffusers/pull/10216/commits/dbc84273f1d2a4124ea61334eb4eb8122b9943e4.

sayakpaul commented on tests/lora/utils.py:
```diff
@@ -1562,7 +1563,6 @@ def test_simple_inference_with_text_denoiser_multi_adapter_weighted(self):
                 "output with no lora and output with lora disabled should give same results",
             )
 
-    @skip_mps
```
Why are we removing this?

hameerabbasi replied:
I added it earlier, assuming it was an MPS limitation. I was wrong, it was a legitimate failure.

sayakpaul replied:
Then we should keep the `skip_mps` decorator right. Why the removal?

hameerabbasi replied:
It passes on MPS, I added it earlier in this PR based on an incorrect assumption that it was MPS causing the failure.

hameerabbasi replied:
Whoops, I wasn't the one to add it, but it does pass on MPS for AuraFlow at least.

Done in https://github.com/huggingface/diffusers/pull/10216/commits/fb5f5f7f24f4511a5ce76495149298f623fd6790.

sayakpaul commented on tests/lora/utils.py:
```diff
@@ -1594,17 +1593,26 @@ def test_lora_fuse_nan(self):
+                    filtered_tower_names = [
+                        tower_name for tower_name in possible_tower_names if hasattr(pipe.transformer, tower_name)
+                    ]
+                    if len(filtered_tower_names) == 0:
+                        pytest.xfail(
+                            reason=f"`pipe.transformer` didn't have any of the following attributes: {possible_tower_names}."
+                        )
```
We can safely remove these. We don't really use `pytest.xfail` inside a test like this.

hameerabbasi replied:
Done in https://github.com/huggingface/diffusers/pull/10216/commits/d2ecef4b735d0d9b09e2067482dd223ec23e6c4b.

hameerabbasi commented:
@sayakpaul Last CI ping. 😅

hameerabbasi commented:
@sayakpaul CI blocked again. 🥹

hameerabbasi commented:
@sayakpaul CI needs approval. 😅

hameerabbasi commented:
@sayakpaul CI ping. 🥹

hameerabbasi commented:
@sayakpaul Since I'm a first-time contributor, you'll need to approve the workflows on GitHub, opening/closing the PR or pushing commits doesn't seem to work. GitHub shows me the following notice on desktop, but the corresponding notice is invisible on mobile:
![grafik](https://github.com/user-attachments/assets/dac93007-67be-43ce-87d6-067bc16d930d)

hameerabbasi commented:
@sayakpaul Another approval on CI please.

hameerabbasi commented:
@sayakpaul Merge before other issues pop up? 😛

AstraliteHeart commented:
@sayakpaul can we please get this merged? very excited about getting AF ecosystem up to speed.

hameerabbasi commented:
@sayakpaul The CI failures seem unrelated:

```
FAILED tests/pipelines/test_pipelines.py::DownloadTests::test_local_save_load_index - huggingface_hub.errors.LocalEntryNotFoundError
```

sayakpaul commented:
Thanks for your contributions.

hameerabbasi commented:
Thanks so much for the reviews and patience!
